normalization should not change KeyDataPair to NameValuePair

thuri opened the pull request:
Fixes #532 
See https://github.com/HtmlUnit/htmlunit/commit/39c58d2b114bebb157be57d63416f629630eee07#r93713845

thuri committed:
normalization should not change KeyDataPair to NameValuePair Fixes #532

rbri commented on src/main/java/com/gargoylesoftware/htmlunit/util/NameValuePair.java:
```diff
@@ -90,4 +90,8 @@ public int hashCode() {
         return name_ + "=" + value_;
     }
+
+    public NameValuePair normalized() {
```
can you please add the jdoc for this method including the INTERNAL API hint

thuri replied:
done. Squashed and force pushed to PR source branch

rbri commented on src/test/java/com/gargoylesoftware/htmlunit/WebRequestTest.java:
```diff
@@ -425,4 +428,19 @@ public void getRequestParametersFromUrlEncodedBodyPost() throws Exception {
         assertEquals(0, request.getRequestParameters().size());
     }
+
+    @Test
```
i like to have a bit more detailed test - checking the values of the new normalized version
and maybe also testing the normalization - having a parameter with null here or two separate tests

thuri replied:
I elaborated a bit on the tests. Please have a look at the new commit. If you think i should add some more testing please let me know.

rbri replied:
fine for me

rbri reviewed (changes requested):
Looks reasonable for me, thanks for this.
Please add yourself as author to all changed files and maybe also to the pom (if you are not already in the list)

rbri commented:
Jenkins will have a look at this....

rbri commented:
sorry we are still at java8 :-)

    /var/lib/jenkins/workspace/htmlunit_PR-533/src/test/java/com/gargoylesoftware/htmlunit/WebRequestTest.java:[437,42] cannot find symbol
    symbol:   method of(com.gargoylesoftware.htmlunit.util.NameValuePair,com.gargoylesoftware.htmlunit.util.KeyDataPair)
    location: interface java.util.List

thuri commented:
Yeah I've seen that but somehow eclipse didn't warn me when i used Java 11 features.
I corrected the problematic part. At least I hope so

rbri commented on src/main/java/com/gargoylesoftware/htmlunit/util/KeyDataPair.java:
```diff
@@ -159,4 +160,25 @@ private void readObject(final ObjectInputStream ois) throws ClassNotFoundExcepti
+    public KeyDataPair normalized() {
+        final KeyDataPair pair = new KeyDataPair(
+            this.getName(),
+            this.fileObject_ != null ? this.fileObject_ : new File(this.getValue()),
```
new File(...) might do some normalization on the file name leading in the end to some new value -> Testcase?
maybe it is better to have a private constructor that allows to pass the values 1:1?

What do you think?

rbri commented:
tests are looking good !

thuri committed:
private copy constructor for KeyDataPair

this should prevent manipulation of fields and make sure that
"normalized" instances are identical copies if the source object is
normalized already

thuri commented:
@rbri  please have a look at the additional commit. Perhaps that is what you had in mind ...

rbri commented:
i have nothing in mind ;-)
thanks a lot - will inform when the new snapshot is availble

and all the best for 2023

thuri commented:
Thanks a lot and the all the best for you too.
